### ISSUE TITLE
Apply spam protection to spec/game join messages

### ISSUE THREAD

**Issue by snailx3** [ddnet/ddnet#1144]
With team_change_delay default value being 3, one can spam spectator => ingame forever, effectively flooding the chat for everyone.. even more efficient with a dummy; some sort of spam protection should be applied to that: either do not display the messages anymore if he's been spamming it, or prevent him from joining game/spectators (the same way you aren't able to send a message anymore if spam protections triggers on you)

**Comment by ChillerDragon** [comment]
Preventing from doing it isn't too nice for tees who use it as selfkill. For example mobile players without a selfkill button. But hiding it is a good idea imo. In DDNet++ we added the possibility to hide join and leave messages completly or for specific names. But it needs an active admin. An automatic detectiom would be good.

**Comment by Learath2** [comment]
Maybe all system messages should go through a ring buffer per person and b e throttled. If too many messages are generated they'd be dropped due to the nature of the ring buffer and we'd have consistent rate limiting.

**Comment by def-** [comment]
Yep, sounds good. But do you want to delay reading out of the ring buffer? Or will it be emptied every tick?

**Comment by ChillerDragon** [comment]
You sure that it is an good idea? For me it sounds like if some troll reconnects or spams me in whisper i loose actual important server messages from people that left/join or tile messages like jetpack/solo/hitothers.

**Comment by heinrich5991** [comment]
@Learath2 suggests throttling joins.

@edg-l proposes moving join messages to a different place.

I think the system messages could go through the normal spam protection subsystem.

**Comment by Learath2** [comment]
> I think the system messages could go through the normal spam protection subsystem.

Only if you are careful with what goes through, spamming yourself is not an issue. Only system messages that are being broadcast, that aren't critical should go through spam protection and it might not be trivial I'm unsure whether the information of who truly generated this message is always available.

**Comment by heinrich5991** [comment]
> that aren't critical

Which player-initiated, broadcast system messages are critical?

**Comment by ChillerDragon** [comment]
> > that aren't critical
> 
> Which player-initiated, broadcast system messages are critical?

Not seeing finishes would be weird.
